### PR TITLE
Tas toolkit issue188 3d visualization

### LINKED ISSUE #188: TAS_Toolkit: Add TBD visualisation
add TBD Visualization

### PR DESCRIPTION
**Pull request by michaldengusiak**
### NOTE: Depends on 

 ### Issues addressed by this PR

 Fixes #188 

 

 ### Test files

[Test File](https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/TAS_Toolkit/TAS_Toolkit-Issue188_Visualisation?csf=1&e=Qjw6B7)

 ### Changelog
- add TBD visualisation

### CODE REVIEW

File: TAS_Engine/Query/ThermostatLimits.cs [hunk 1]
```diff
@@ -64,6 +69,9 @@ public static float GetSingleValueUpperLimit(this TBD.Thermostat tbdICThermostat
             return maxUL;
         }
 
+        [Description("BH.Engine.TAS Query- Get Single Lower Limit")]
```
Comment (FraserGreenroyd) [inline TAS_Engine/Query/ThermostatLimits.cs]: Please remove BH.Engine.TAS Query as discussed :smile:

File: TAS_Engine/Query/ThermostatLimits.cs [hunk 2]
```diff
@@ -25,11 +25,16 @@
...
 using System.Text;
 using System.Threading.Tasks;
+using System.ComponentModel;
+using BH.oM.Reflection.Attributes;
 
 namespace BH.Engine.TAS
 {
     public static partial class Query
     {
+        [Description("BH.Engine.TAS Query- Get Single Upper Limit")]
```
Comment (FraserGreenroyd) [inline TAS_Engine/Query/ThermostatLimits.cs]: Please remove BH.Engine.TAS Query as discussed :smile:

File: TAS_Engine/Query/TBD3dVisualisation.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+using BH.oM.Environment.Elements;
+
+using BH.oM.Environment.Fragments;
+using System;
+
+namespace BH.Engine.TAS
+{
+    public static partial class Query
+    {
+        [Description("BH.Engine.TAS Query- Get TAS TBD 3d Visulisation, when finish close window and click False to end process. Use Shift + Middle mouse")]
```
Comment (FraserGreenroyd) [inline TAS_Engine/Query/TBD3dVisualisation.cs]: Please remove BH.Engine.TAS Query as discussed :smile:

File: TAS_Engine/Query/TBD3dVisualisation.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+            if (string.IsNullOrEmpty(tbdFile))
+            {
+                BH.Engine.Reflection.Compute.RecordError("Please provide a valid TBD file path");
+                return false;
+            }
+
+            TBDDocument tbdDocument = new TBDDocument();
+            tbdDocument.openReadOnly(tbdFile);
+            tbdDocument.ShowVisualisation();
```
Comment (FraserGreenroyd) [inline TAS_Engine/Query/TBD3dVisualisation.cs]: Check that the `tbdDocument` is not equal to `null` first?

File: TAS_Engine/Query/BuildingElement.cs [hunk 5]
```diff
@@ -29,14 +29,19 @@
...
+using BH.oM.Reflection.Attributes;
 
 namespace BH.Engine.TAS
 {
     public static partial class Query
     {
 
         /***************************************************/
-
+        [Description("BH.Engine.TAS Query- Get TBD Building Element from building by Name")]
```
Comment (FraserGreenroyd) [inline TAS_Engine/Query/BuildingElement.cs]: Please remove BH.Engine.TAS Query as discussed :smile:

File: TAS_Engine/Compute/TBDSizing.cs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                BH.Engine.Reflection.Compute.RecordError("Please provide a valid TBD file path");
+                return false;
+            }
+
+            TBDDocument tbdDocument = new TBDDocument();
+            tbdDocument.open(tbdFile);
+
+            string tsdFile = tbdFile.Replace(".tbd", ".tsd");
+
+            tbdDocument.simulate(startDay, endDay, autoViewResults ? 1 : 0, 1, 0, 0, tsdFile, 0, 0);
```
Comment (FraserGreenroyd) [inline TAS_Engine/Compute/TBDSizing.cs]: Check the `tbdDocument` is not equal to `null` first?

File: TAS_Engine/Compute/TBDSizing.cs [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+            {
+                intrefcountTBD = Marshal.FinalReleaseComObject(tbdDocument);
+            } while (intrefcountTBD > 0);
+            tbdDocument = null;
+
+            return true;
+
+        }
+
+        [Description("BH.Engine.TAS Compute - Run Dynamic Simulation")]
```
Comment (FraserGreenroyd) [inline TAS_Engine/Compute/TBDSizing.cs]: Please remove BH.Engine.TAS Query as discussed :smile:

File: TAS_Engine/Compute/TBDSizing.cs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+using BH.oM.Environment.Fragments;
+using System;
+
+namespace BH.Engine.TAS
+{
+    public static partial class Compute
+    {
+
+        [Description("BH.Engine.TAS Compute - Run Design Day Sizing")]
```
Comment (FraserGreenroyd) [inline TAS_Engine/Compute/TBDSizing.cs]: Please remove BH.Engine.TAS Query as discussed :smile:

File: TAS_Engine/Query/ThermostatLimits.cs (hunk 1 above)
Comment (michaldengusiak) [reply TAS_Engine/Query/ThermostatLimits.cs]: done

File: TAS_Engine/Query/ThermostatLimits.cs (hunk 2 above)
Comment (michaldengusiak) [reply TAS_Engine/Query/ThermostatLimits.cs]: done

File: TAS_Engine/Query/TBD3dVisualisation.cs (hunk 3 above)
Comment (michaldengusiak) [reply TAS_Engine/Query/TBD3dVisualisation.cs]: done

File: TAS_Engine/Query/TBD3dVisualisation.cs (hunk 4 above)
Comment (michaldengusiak) [reply TAS_Engine/Query/TBD3dVisualisation.cs]: done

File: TAS_Engine/Query/BuildingElement.cs (hunk 5 above)
Comment (michaldengusiak) [reply TAS_Engine/Query/BuildingElement.cs]: done

File: TAS_Engine/Compute/TBDSizing.cs (hunk 6 above)
Comment (michaldengusiak) [reply TAS_Engine/Compute/TBDSizing.cs]: done

File: TAS_Engine/Compute/TBDSizing.cs (hunk 7 above)
Comment (michaldengusiak) [reply TAS_Engine/Compute/TBDSizing.cs]: done

File: TAS_Engine/Compute/TBDSizing.cs (hunk 8 above)
Comment (michaldengusiak) [reply TAS_Engine/Compute/TBDSizing.cs]: done

**Pull request merged**